promql/engine.go

```diff
@@ -1325,16 +1395,100 @@ func (ev *evaluator) eval(expr parser.Expr) (parser.Value, storage.Warnings) {
 			newEv.interval = ev.noStepSubqueryIntervalFn(rangeMillis)
 		}
 
-		// Start with the first timestamp after (ev.startTimestamp - offset - range)
+		subqueryEvalTime := timestamp.FromFloatSeconds(e.Timestamp)
+		subqueryStartTime := ev.startTimestamp - offsetMillis - rangeMillis
+		if subqueryEvalTime != 0 {
+			newEv.endTimestamp = subqueryEvalTime - offsetMillis
+			subqueryStartTime = newEv.endTimestamp - rangeMillis
+		}
+
+		// Start with the first timestamp after subqueryStartTime
 		// that is aligned with the step (multiple of 'newEv.interval').
-		newEv.startTimestamp = newEv.interval * ((ev.startTimestamp - offsetMillis - rangeMillis) / newEv.interval)
-		if newEv.startTimestamp < (ev.startTimestamp - offsetMillis - rangeMillis) {
+		newEv.startTimestamp = newEv.interval * (subqueryStartTime / newEv.interval)
+		if newEv.startTimestamp < subqueryStartTime {
 			newEv.startTimestamp += newEv.interval
 		}
 
 		res, ws := newEv.eval(e.Expr)
 		ev.currentSamples = newEv.currentSamples
 		return res, ws
+	case *parser.StepInvariantExpr:
+		switch ce := e.Expr.(type) {
+		case *parser.StringLiteral, *parser.NumberLiteral:
+			return ev.eval(ce)
+		}
+
+		if e.Result != nil {
+			return e.Result, nil
+		}
+		newEv := &evaluator{
+			startTimestamp:           ev.startTimestamp,
+			endTimestamp:             ev.startTimestamp,
+			interval:                 ev.interval,
+			ctx:                      ev.ctx,
+			currentSamples:           ev.currentSamples,
+			maxSamples:               ev.maxSamples,
+			logger:                   ev.logger,
+			lookbackDelta:            ev.lookbackDelta,
+			noStepSubqueryIntervalFn: ev.noStepSubqueryIntervalFn,
+		}
+
+		canExtrapolate := canExtrapolateStepInvariantExpr(e.Expr)
+		if !canExtrapolate {
+			newEv.endTimestamp = ev.endTimestamp
+		}
+
+		res, ws := newEv.eval(e.Expr)
+		ev.currentSamples = newEv.currentSamples
+
+		_, isSubquery := e.Expr.(*parser.SubqueryExpr)
+		if canExtrapolate && !isSubquery {
+			switch val := res.(type) {
+			case Matrix:
+				for i := range val {
+					if len(val[i].Points) != 1 {
```

Comment:
Why can a matrix only have exactly one value per series?

Reply:
Since we exclude matrix selector and subqueries from this duplication here, then there is no other node in the PromQL that can return a matrix with multiple samples.

Reply:
I'm confused, what code path lands us here?

Reply:
> I'm confused, what code path lands us here?

Does the look clear now with recent changes? Since we return above for matrix selector and subquery, here we must have exactly 1 sample.